biogeophys/FatesHydroWTFMod.F90

```diff
@@ -190,6 +205,90 @@ module FatesHydroWTFMod
+  ! of numerical integration.
+  ! ============================================================================
+  
+  subroutine set_min_max(this,th_res,th_sat)
+
+      ! This routine uses max_sf_interp and min_sft_interp
+      ! to define the bounds of where the linear ranges start and stop
+      
+      class(wrf_type)   :: this
+      real(r8),intent(in) :: th_res
+      real(r8),intent(in) :: th_sat
+      
+      this%th_max      = max_sf_interp*(th_sat-th_res)+th_res
+      this%th_min      = min_sf_interp*(th_sat-th_res)+th_res
+      this%psi_max     = this%psi_from_th(this%th_max-tiny(this%th_max))
+      this%dpsidth_max = this%dpsidth_from_th(this%th_max-tiny(this%th_max))
+      this%psi_min     = this%psi_from_th(this%th_min+tiny(this%th_min))
+      this%dpsidth_min = this%dpsidth_from_th(this%th_min+tiny(this%th_min))
+      
+  end subroutine set_min_max
+      
+  ! ============================================================================
+  
+  function psi_linear_res(this,th) result(psi)
+
+      ! Calculate psi in linear range below residual
+      
+      class(wrf_type)   :: this
+      real(r8),intent(in)  :: th    ! vol. wat. cont   [m3/m3]
+      real(r8)             :: psi   ! Matric potential [MPa]
+      
+      psi = this%psi_min + this%dpsidth_min * (th-this%th_min)
+      
+  end function psi_linear_res
+  
+  ! ===========================================================================
+  
+  function psi_linear_sat(this,th) result(psi)
+
+      ! Calculate psi in linear range above saturation
+      
+      class(wrf_type)   :: this
+      real(r8),intent(in)  :: th    ! vol. wat. cont   [m3/m3]
+      real(r8)             :: psi   ! Matric potential [MPa]
+            
+      psi = this%psi_max + this%dpsidth_max * (th-this%th_max) 
+      
+  end function psi_linear_sat
+
+  ! ===========================================================================
+  
+  function th_linear_sat(this,psi) result(th)
+
+      ! Calculate th from psi in linear range above saturation
+      
+      class(wrf_type)   :: this
+      real(r8),intent(in)   :: psi   ! Matric potential [MPa]
+      real(r8)              :: th    ! vol. wat. cont   [m3/m3]
+     
+      th = this%th_max + (psi-this%psi_max)/this%dpsidth_max
+
+  end function th_linear_sat
+
+  ! ===========================================================================
+  
+  function th_linear_res(this,psi) result(th)
+
+      ! Calculate th from psi in linear range above saturation
+      
+      class(wrf_type)   :: this
+      real(r8),intent(in)   :: psi   ! Matric potential [MPa]
+      real(r8)              :: th    ! vol. wat. cont   [m3/m3]
+     
+      th = this%th_min + (psi-this%psi_min)/this%dpsidth_min
+
+  end function th_linear_res
+  
+  ! ===========================================================================
+  
   subroutine set_wrf_param_base(this,params_in)
     class(wrf_type)     :: this
     real(r8),intent(in) :: params_in(:)
```

Comment:
The variables change depending on the hypothesis we choose, so each variable would have an index for each set_wrf_param function.  In the specific set_wrf_param functions, the array indices go immediately to their variables, so there is no ambiguity about the array indices and which variable they connect to.  

Here are the three different locations for setting the parameters (variables) of each specific hypothesis, I'm not against using named indices, but it might be a lot of additional code.   For the three specific hypotheses, that's 24 total variables, and some of them will have subtly different names, like theta sat for vg, or theta sat for tfs, etc:

https://github.com/NGEET/fates/blob/675aa9555be6fcd3a653f60ebd019c1e21fcdd0b/biogeophys/FatesHydroWTFMod.F90#L363-L392
https://github.com/NGEET/fates/blob/675aa9555be6fcd3a653f60ebd019c1e21fcdd0b/biogeophys/FatesHydroWTFMod.F90#L620-L653
https://github.com/NGEET/fates/blob/675aa9555be6fcd3a653f60ebd019c1e21fcdd0b/biogeophys/FatesHydroWTFMod.F90#L756-L785

Should I go ahead and add them  @xuchongang , or hold off on them for now?

Reply:
@rgknox , I think I am fine with the current formulation as it provide flexibility.